### PR TITLE
feat!: EigenStepper change internally to FreeVector

### PR DESCRIPTION
**Pull request by asalzburger**
This PR changes the internal representation of parameters in the EigenStepper to a `FreeVector`.
The covariance matrix is already expressed as a `FreeMatrix`, hence this is a logical internal choice.

BREAKING CHANGE: The member structure of the Steppers is (still) publicly exposed and hence has to be regarded as a public interface.

### CODE REVIEW

**Comment by msmk0** [comment]
@asalzburger I think the title should be `refactor: ...`. `chore: ...` is for infrastructure.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (FabianKlimpel) [inline Core/include/Acts/Propagator/EigenStepper.hpp:108]:
```suggestion
    int q = 1.;
```
Historically it was convenient to store it as a `double` since we need `q/p` calculated all the time. If you store `q/p` directly then we can change it to `int` (assuming that we never propagate quarks), right?

File: Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp
Comment (FabianKlimpel) [inline Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp:179]: This line suggests that the charge is -1,0,1. If this is the (reasonable) way to go, then the charge could even become a bool. The sign can be obtained from the `pars[eFreeQOverP]`

File: Tests/UnitTests/Core/Propagator/StepperTests.cpp
Comment (FabianKlimpel) [inline Tests/UnitTests/Core/Propagator/StepperTests.cpp:209]: Why are tests for momentum and direction missing?

File: Tests/UnitTests/Core/Propagator/StepperTests.cpp
Comment (FabianKlimpel) [inline Tests/UnitTests/Core/Propagator/StepperTests.cpp:428]: This keeps the direction unchanged

File: Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp:179]: Sorry, this will be updated to use the charge ...

File: Tests/UnitTests/Core/Propagator/StepperTests.cpp
Comment (asalzburger) [reply Tests/UnitTests/Core/Propagator/StepperTests.cpp:428]: This test was stupid anyway, it was testing that if I created a non-well-defined freeParams with a non-normalised direction the `State` would do the normalization. I think we should remove this test and do garbage-in-garbage-out.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:108]: I decided to keep it consistent with the return value of `parameters.charge()`

File: Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp:179]: Updated.

File: Tests/UnitTests/Core/Propagator/StepperTests.cpp
Comment (asalzburger) [reply Tests/UnitTests/Core/Propagator/StepperTests.cpp:209]: RE-introduced.

**Review by asalzburger — COMMENTED** [review]
Adressed PR comments.

**Review by FabianKlimpel — APPROVED** [review]

**Comment by asalzburger** [comment]
Unit test is failing because of equality checking of vectors and not `CHECK_CLOSE_ABS`. Will change that.

File: Tests/UnitTests/Core/Propagator/StepperTests.cpp
Comment (FabianKlimpel) [inline Tests/UnitTests/Core/Propagator/StepperTests.cpp:211]:
```suggestion
  CHECK_CLOSE_ABS(es.position(esState), pos, eps);
  CHECK_CLOSE_ABS(es.direction(esState), dir, eps;
  CHECK_CLOSE_ABS(es.momentum(esState), absMom, eps);
  CHECK_CLOSE_ABS(es.charge(esState), charge, eps);
  CHECK_CLOSE_ABS(es.time(esState), time, eps);
```

**Review by FabianKlimpel — CHANGES_REQUESTED** [review]

**Review by msmk0 — CHANGES_REQUESTED** [review]
@asalzburger The code changes look good. I have some questions regarding the q/p interpretation that we should address before merging this in.

**Comment by msmk0** [comment]
@asalzburger @FabianKlimpel One more comment: this changes the public interface in a module that we have explicitly defined as [public](https://acts.readthedocs.io/en/latest/versioning.html). I agree that is is logically an internal interface (and will be also technically be one once the members become private), but is it not at the moment. We should stick to our own rules and mark this PR as breaking, i.e. `refactor!`.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:197]: Good idea, actually - will change that.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:211]: I totally agree.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:197]: The concept checks against a return type, and auto can not be used for the moment, I will have to check.

Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:197]: I move that to 'future:' as well.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:91]: > We have to be a bit careful here in what we want to actually use this for: if this should track the curvature then this is fine as you automatically get zero curvature for neutral particles. This would correctly give you the right trajectory but you loose the momentum information.
> 
> In the track parameters, this parameter is either `q/p` or `1/p` for uncharged particles. If we take the same definition here, than we can store the existing q/p values from the parameters.
> 
> My personal take is that the q/p parameter in the internal free vector should represent the curvature-defining parameter, i.e. its zero for neutral parameters and different from the q/p parameter in the track parameters. Then the actual computations do not need to care whether we have a charged or neutral particle. But that would mean that we still need to track `q` and `p` separately and update `p` after every propagation step. This only really applies to the dense stepper anyways.
> 
> Btw. no such complication exists for the straight line stepper. There the internal q/p parameter should be the same as the q/p parameter in the track parameters.

Mmmm, if you ask neutral bound parameters to be converted into free bound parameters then are you expecting `parameters[eOverP]` to be 0?

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (msmk0) [reply Core/include/Acts/Propagator/EigenStepper.hpp:197]: I recently updated the `TrackParametersConcept` to require only convertibility not identity. That is relatively straight-forward but we can do that at a later stage.

Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:197]: Ah, but that's good to know!

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (msmk0) [reply Core/include/Acts/Propagator/EigenStepper.hpp:91]: It depends. For the pure track parameters: no. For neutral track parameters the `q/p` parameter will hold `1/p` and will be non-zero for finite p. This is clearly outlined e.g. in the charge interpreter documentation in `EventData/Charge.hpp`.

Here, the `FreeVector` is a purely internal state that does **not** have to have the identical content as the `FreeVector` in the parameters. Only when we request the track parameters should it match the track parameters definition. I am saying that for the particular, internal use within the `EigenStepper` it might be beneficial to use a slightly different definition (as outlined above) for this parameter.

Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:91]: That would however re-introduce a conversion in all conversion function, which now can conveniently be called with 
`freeToBound(...,state.pars,...)`, which I think it's not worth it.

The `FreeVector` will be a private member of the `State` and I think we can go ahead and chose the most convenient version for us, even if it has a similar convention as mentioned above. Besides, we still have an open question whether reordering the internal parameters would speed up the covariance transport (could not reproduce so easily now).

Comment (msmk0) [reply Core/include/Acts/Propagator/EigenStepper.hpp:91]: > That would however re-introduce a conversion in all conversion function, which now can conveniently be called with
> `freeToBound(...,state.pars,...)`, which I think it's not worth it.

That is a good point. From that perspective it is much easier to use exactly the same definition as in the track parameters. However, than I would suggest to not compute it manually as `par.charge() / par.absoluteMomentum()` but to use the existing `q/p` value from the input parameters. The track parameters already take care of using either `q/p` or `1/p` consistently.

Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:91]: Excellent point, will change that.

Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:91]: ... but only when we have fully introduced the FreeParameters, though, because at the moment, the constructor is templated on the `parameters_t` type, and hence I have no access to either `eFreeQoverP` or `eBoundQoverP`, I plan to split this into two constructors anyway because for the free parameters the constructor would just be a assignment operation.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:197]: Fill do that with the Concept rework when the member methods are all defined.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:108]: Fixed.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:210]: Fill do that with the Concept rework.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (FabianKlimpel) [reply Core/include/Acts/Propagator/EigenStepper.hpp:90]: > ... but only when we have fully introduced the FreeParameters, though, because at the moment, the constructor is templated on the `parameters_t` type, and hence I have no access to either `eFreeQoverP` or `eBoundQoverP`, I plan to split this into two constructors anyway because for the free parameters the constructor would just be a assignment operation.

Which is exactly what I did in #86

Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:90]: I see.

Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp]: I changed it and enabled for the moment only construction from `BoundParameters` as @FabianKlimpel has suggested to do.

**Review by asalzburger — COMMENTED** [review]
I think, I addressed what I can do at the moment with this PR, I will change the StraightLineStepper accordingly.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (FabianKlimpel) [inline Core/include/Acts/Propagator/EigenStepper.hpp:76]: When I added these variables to the parameters, the naming pattern for variables weren't developed yet. I think it has to be something like `s_isBoundParameters` or so. @msmk0 could help suggesting names.

Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:76]: Good call ...

Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:76]: I renamed it to s_boundRepresentation

**Review by FabianKlimpel — APPROVED** [review]

**Review by asalzburger — COMMENTED** [review]
Ok, waiting for CI to run through @msmk0 - are you ok with these changes?

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (msmk0) [inline Core/include/Acts/Propagator/EigenStepper.hpp:81]:
```suggestion
    template <typename charge_t>
    explicit State(std::reference_wrapper<const GeometryContext> gctx,
                   std::reference_wrapper<const MagneticFieldContext> mctx,
                   const SingleBoundTrackParameters<charge_t>& par, NavigationDirection ndir = forward,
                   double ssize = std::numeric_limits<double>::max(),
                   double stolerance = s_onSurfaceTolerance)
```
Whether a track parameter is bound or free is already encoded in its type name. I do not really see why we need to add an extra member variable here to disambiguate. By using the concrete type (with just the charge template), we can let the compiler do the selection work for us and we do not have to rely on the `std::enable_if_t` magic (which I personally always find a bit hard to follow).

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (msmk0) [inline Core/include/Acts/Propagator/EigenStepper.hpp:108]: I thought we decided to use `double` here to be consistent?

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (msmk0) [inline Core/include/Acts/Propagator/EigenStepper.hpp:212]: This should also take the charge into account. Otherwise, you get wrong values e.g. for doubly-charge ions. See the documentation/implementation in `Charge.hpp`.

File: Core/include/Acts/EventData/SingleBoundTrackParameters.hpp
Comment (msmk0) [inline Core/include/Acts/EventData/SingleBoundTrackParameters.hpp:38]: I think this is not needed. See below.

File: Core/include/Acts/EventData/SingleFreeTrackParameters.hpp
Comment (msmk0) [inline Core/include/Acts/EventData/SingleFreeTrackParameters.hpp:31]: I think this is not needed. See below.

**Review by msmk0 — CHANGES_REQUESTED** [review]
@asalzburger I added some suggestions on how to simplify the type selection. There is also some ambiguity when converting from `q/p` to `p` since you ignore `q`.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp]: Dropping this following @msmk0 's suggestion.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:81]: Agreed.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:108]: Oh yes, will revert this.

File: Core/include/Acts/Propagator/EigenStepper.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp]: Done.

**Review by msmk0 — APPROVED** [review]

**Comment by asalzburger** [comment]
Cool, I will update the StraighLineStepper then accordingly.

**Comment by paulgessinger** [comment]
So the reason this is breaking is because previously we (implicitly) considered members of the state as public interface, and now we don't? 

Can you formulate the BREAKING CHANGE trailer so that this becomes clear in isolation?

**Comment by msmk0** [comment]
@paulgessinger Almost. Making the members private will be a another change. Here, the (still public) members of the state struct are modified and thus the public interface changes.

**Pull request merged**
